Add documentation for plugin testing (#401)
cloudoperators/greenhouse: docs/user-guides/plugin/plugin-tests.md

---
title: "Testing a Plugin"
linkTitle: "Testing a Plugin"
landingSectionIndex: false
weight: 2
description: >
  Guidelines for testing plugins contributed to the Greenhouse project.
---

## Plugin Testing Requirements

> All plugins contributed to this repository should include comprehensive [Helm Chart Tests](https://helm.sh/docs/topics/chart_tests/) using the `bats/bats-detik` testing framework in conjunction with standard Helm Chart Tests. This ensures our plugins are robust, deployable, and catch potential issues early in the development cycle.

Author: That's a great catch. I have fixed in this commit https://github.com/cloudoperators/greenhouse/pull/401/commits/0f318577ea9b862c854aa5cf6f4d994f1bd057eb.

**What is bats/bats-detik?**

The [bats/bats-detik](https://github.com/bats-core/bats-detik) framework simplifies end-to-end (e2e) Testing in Kubernetes. It combines the Bash Automated Testing System (`bats`) with Kubernetes-specific assertions (`detik`). This allows you to write test cases using natural language-like syntax, making your tests easier to read and maintain.

**Implementing Tests**

> 1. Create a **tests** folder inside your Plugin's Helm Chart `templates` folder to store your test resources.

Reviewer: ```suggestion
1. Create a `/tests` folder inside your Plugin's Helm Chart `templates` folder to store your test resources.
```

2. **ConfigMap definition**:

>    - Create a `test-<plugin-name>-config.yaml` file in the `templates/tests` directory to define a ConfigMap that will hold your test script.

Reviewer: ```suggestion
   - Create a `test-<plugin-name>-config.yaml` file in the `templates/tests` directory to define a `ConfigMap` that will hold your test script.
```

>    - This `ConfigMap` contains the test script `run.sh` that will be executed by the test Pod to run your tests.

Reviewer: ```suggestion
   - This `ConfigMap` contains the test script `run.sh` that will be executed by the test `Pod` to run your tests.
```

```yaml
{{- if .Values.testFramework.enabled -}}
apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ .Release.Name }}-test
  namespace: {{ .Release.Namespace }}
  labels:
    type: integration-test
  annotations:
    "helm.sh/hook": test

>     "helm.sh/hook-weight": "-5" # Run before the test pod

Reviewer: ```suggestion
    "helm.sh/hook-weight": "-5" # Installed and upgraded before the test pod
```

    }

    @test "Verify successful creation and bound status of {{ .Release.Name }} persistent volume claims" {
        try "at most 3 times every 5s to get persistentvolumeclaims named '{{ .Release.Name }}.*' and verify that '.status.phase' is 'Bound'"
    }

    @test "Verify successful creation and available replicas of {{ .Release.Name }} Prometheus resource" {
        try "at most 3 times every 5s to get prometheuses named '{{ .Release.Name }}' and verify that '.status.availableReplicas' is more than '0'"
    }

    @test "Verify creation of required custom resource definitions (CRDs) for {{ .Release.Name }}" {
        verify "there is 1 customresourcedefinition named 'prometheuses'"
        verify "there is 1 customresourcedefinition named 'podmonitors'"
    }
{{- end -}}
```

> **Note:** You can use [this guide](https://github.com/bats-core/bats-detik/blob/master/examples/bats/test_kubectl_and_oc.sh) for reference when writing your test assertions.

3. **Test Pod Definition**:

   - Create a `test-<plugin-name>.yaml` file in the `templates/tests` directory to define a `Pod` that will run your tests.

>    - This test `Pod` will mount the `ConfigMap` created in the previous step and will execute the test script `run.sh`.

Author: I followed the official [helm Chart Test](https://helm.sh/docs/topics/chart_tests/) documentation and they are using `kind: Pod` with `restartPolicy: Never`. The cleanup of the Helm Chart Test Resources will be handled by the Helm directly if the annotation `"helm.sh/hook-delete-policy": "before-hook-creation,hook-succeeded"` exist in the test manifests.

 [link to reference](https://helm.sh/docs/topics/charts_hooks/#hook-deletion-policies)

>    - This test `Pod` will mount the `ConfigMap` created in the previous step and will execute the test script `run.sh`.

Reviewer 2: Ok, just saw that the docs mention jobs while providing a Pod as an example. While it is possible to use jobs for hooks with helm3, retrieving the logs for the test only works with Pods.
There are several bugs reported that `helm test --logs` results in errors if any resources other than a Pod are used as hooks. 🫠

>    - This test `Pod` will mount the `ConfigMap` created in the previous step and will execute the test script `run.sh`.

Author: Thank you for the information, Ivo! I did not know about it till now :)

```yaml
 {{- if .Values.testFramework.enabled -}}
 apiVersion: v1
 kind: Pod
 metadata:
   name: {{ .Release.Name }}-test
   namespace: {{ .Release.Namespace }}
   labels:
     type: integration-test
   annotations:
     "helm.sh/hook": test
     "helm.sh/hook-delete-policy": "before-hook-creation,hook-succeeded"
 spec:
   serviceAccountName: {{ .Release.Name }}-test
   containers:
     - name: bats-test
       image: "{{ .Values.testFramework.image.registry}}/{{ .Values.testFramework.image.repository}}:{{ .Values.testFramework.image.tag }}"
       imagePullPolicy: {{ .Values.testFramework.image.pullPolicy }}
       command: ["bats", "-t", "/tests/run.sh"]
       volumeMounts:
         - name: tests
           mountPath: /tests
           readOnly: true
   volumes:
     - name: tests
       configMap:
         name: {{ .Release.Name }}-test
   restartPolicy: Never
 {{- end -}}
```

4. **RBAC Permissions**:

> - Create a `test-permissions.yaml` file in the `templates/tests` folder to define the `ServiceAccount` and necessary RBAC permissions for the test Pod.

Reviewer: ```suggestion
- Create the necessary RBAC resources in the `templates/tests` folder with a dedicated `ServiceAccount` and role authorisations so that the test `Pod` can cover test the cases. 
```

- You can use [test-permissions.yaml](https://github.com/cloudoperators/greenhouse-extensions/blob/main/kube-monitoring/charts/templates/tests/test-permissions.yaml) from the `kube-monitoring` as a reference to configure RBAC permissions for your test Pod.

5. **Configure the Test Framework in Plugin's `values.yaml`**:
   - Add the following configuration to your plugin's `values.yaml` file:

```yaml
testFramework:
  enabled: true
  image:
    registry: ghcr.io
    repository: cloudoperators/greenhouse-extensions-integration-test
    tag: main
  imagePullPolicy: IfNotPresent
```

**Contribution Checklist**

Before submitting a pull request:

> - Ensure your plugin's Helm chart includes a `tests` directory.

Reviewer: ```suggestion
- Ensure your plugin's Helm chart includes a `/tests` directory.
```
